### PR TITLE
Train Controlnet new hw support

### PR DESCRIPTION
**Pull request by congw729**
Acceptable training result in 1k dataset.

### CODE REVIEW

File: examples/stable_diffusion_v2/ldm/modules/diffusionmodules/openaimodel.py [hunk 1]
```diff
@@ -154,14 +154,14 @@ def __init__(
             self.h_upd = self.x_upd = self.identity
 
         self.emb_layers = nn.SequentialCell(
-            nn.SiLU().to_float(self.dtype),
+            nn.SiLU().to_float(ms.float32),
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/ldm/modules/diffusionmodules/openaimodel.py:157]: can influence the performance of other SD models. Need to check whether using fp16 is also fine.

File: examples/stable_diffusion_v2/ldm/modules/encoders/text_encoder.py [hunk 2]
```diff
@@ -86,7 +88,7 @@ def __init__(self, d_model, n_head, attn_mask, epsilon, use_quick_gelu, dtype=ms
 
         self.c_proj = nn.Dense(d_model * 4, d_model).to_float(dtype)
         self.mlp = nn.SequentialCell([self.c_fc, self.gelu, self.c_proj])
-        self.ln_2 = LayerNorm([d_model], epsilon=epsilon).to_float(dtype)
+        self.ln_2 = LayerNorm([d_model], epsilon=epsilon).to_float(ms.float32)
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/ldm/modules/encoders/text_encoder.py:91]: same here

File: examples/stable_diffusion_v2/ldm/modules/encoders/text_encoder.py [hunk 3]
```diff
@@ -145,7 +147,7 @@ def __init__(
         self.positional_embedding = Parameter(
             initializer(TruncatedNormal(0.01), [context_length, width], dtype=self.dtype)
         )
-        self.ln_final = LayerNorm([self.width]).to_float(self.dtype)
+        self.ln_final = LayerNorm([self.width]).to_float(ms.float32)
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/ldm/modules/encoders/text_encoder.py:150]: same here

File: examples/stable_diffusion_v2/ldm/modules/encoders/text_encoder.py [hunk 4]
```diff
@@ -188,12 +190,12 @@ def __init__(
     ):
         super().__init__()
         self.dtype = dtype
-        self.ln_1 = LayerNorm((d_model,), epsilon=epsilon).to_float(self.dtype)
+        self.ln_1 = LayerNorm((d_model,), epsilon=epsilon).to_float(ms.float32)
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/ldm/modules/encoders/text_encoder.py:193]: +1

File: examples/stable_diffusion_v2/ldm/modules/encoders/text_encoder.py [hunk 5]
```diff
@@ -188,12 +190,12 @@ def __init__(
...
         self.attn = MultiheadAttention(d_model, n_head, dtype=dtype)
         if is_cross_attention:
-            self.ln_1_kv = LayerNorm((d_model,), epsilon=epsilon).to_float(self.dtype)
+            self.ln_1_kv = LayerNorm((d_model,), epsilon=epsilon).to_float(ms.float32)
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/ldm/modules/encoders/text_encoder.py:196]: +1

File: examples/stable_diffusion_v2/ldm/modules/encoders/text_encoder.py [hunk 6]
```diff
@@ -244,7 +246,7 @@ def __init__(
         self.positional_embedding = Parameter(
             initializer(TruncatedNormal(0.01), [context_length, width], dtype=self.dtype)
         )
-        self.ln_final = nn.LayerNorm((self.width,), epsilon=epsilon).to_float(self.dtype)
+        self.ln_final = nn.LayerNorm((self.width,), epsilon=epsilon).to_float(ms.float32)
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/ldm/modules/encoders/text_encoder.py:249]: +1

File: examples/stable_diffusion_v2/ldm/modules/encoders/text_encoder.py [hunk 7]
```diff
@@ -32,9 +32,11 @@ def __init__(self, d_model, n_head, dtype=ms.float32):
         self.out_proj = nn.Dense(self.embed_dim, self.embed_dim).to_float(dtype)
         self.split = ops.Split(-1, 3)
         self.expand_dims = ops.ExpandDims()
-        self.softmax = nn.Softmax(-1)
+        # self.softmax = nn.Softmax(-1).to_float(ms.float32)
+        self.softmax = nn.Softmax(-1).to_float(dtype)
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/ldm/modules/encoders/text_encoder.py:36]: keep nn.Softmax(-1)

File: examples/stable_diffusion_v2/ldm/modules/diffusionmodules/openaimodel.py [hunk 8]
```diff
@@ -374,7 +374,7 @@ def __init__(
         time_embed_dim = model_channels * 4
         self.time_embed = nn.SequentialCell(
             linear(model_channels, time_embed_dim, dtype=self.dtype),
-            nn.SiLU().to_float(self.dtype),
+            nn.SiLU().to_float(ms.float32),
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/ldm/modules/diffusionmodules/openaimodel.py:377]: +1

File: examples/stable_diffusion_v2/ldm/modules/diffusionmodules/openaimodel.py [hunk 9]
```diff
@@ -386,7 +386,7 @@ def __init__(
                 self.label_emb = nn.SequentialCell(
                     nn.SequentialCell(
                         linear(adm_in_channels, time_embed_dim, dtype=self.dtype),
-                        nn.SiLU().to_float(self.dtype),
+                        nn.SiLU().to_float(ms.float32),
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/ldm/modules/diffusionmodules/openaimodel.py:389]: +1

File: examples/stable_diffusion_v2/ldm/modules/diffusionmodules/openaimodel.py [hunk 10]
```diff
@@ -624,7 +624,7 @@ def __init__(
 
         self.out = nn.SequentialCell(
             normalization(ch),
-            nn.SiLU().to_float(self.dtype),
+            nn.SiLU().to_float(ms.float32),
```
Comment (SamitHuang) [inline examples/stable_diffusion_v2/ldm/modules/diffusionmodules/openaimodel.py:627]: +1

File: examples/stable_diffusion_v2/controlnet_train.md
Comment (SamitHuang) [inline examples/stable_diffusion_v2/controlnet_train.md:1]: Please merge this content in this file into `controlnet.md`

File: examples/stable_diffusion_v2/cldm/cldm.py [hunk 11]
```diff
@@ -24,60 +23,47 @@
...
+
+        # add controlnet init
         self.controlnet = instantiate_from_config(control_stage_config)
         self.control_scales = (
             [strength * (0.825 ** float(12 - i)) for i in range(13)] if guess_mode else ([strength] * 13)
         )
 
+        for param in self.controlnet.get_parameters():
+            print("INFO: controlnet param trainable: ", param.requires_grad)
+            break
```
Comment (hadipash) [inline examples/stable_diffusion_v2/cldm/cldm.py:49]: Remove debugging code

Comment (congw729) [reply examples/stable_diffusion_v2/cldm/cldm.py:49]: Hi. I would change it to `logging.info` to satisfy the potential needs.

File: examples/stable_diffusion_v2/controlnet_train.md
Comment (congw729) [reply examples/stable_diffusion_v2/controlnet_train.md:1]: done

File: examples/stable_diffusion_v2/cldm/cldm.py (hunk 11 above)
Comment (hadipash) [reply examples/stable_diffusion_v2/cldm/cldm.py]: I meant the `break` statement.

Comment (congw729) [reply examples/stable_diffusion_v2/cldm/cldm.py]: Oh. Thank you. Done.

File: examples/stable_diffusion_v2/train_text_to_image.py [hunk 12]
```diff
@@ -172,6 +171,15 @@ def parse_args():
     # parser.add_argument("--cond_stage_trainable", default=False, type=str2bool, help="whether text encoder is trainable")
     parser.add_argument("--use_ema", default=False, type=str2bool, help="whether use EMA")
     parser.add_argument("--clip_grad", default=False, type=str2bool, help="whether apply gradient clipping")
+    # parser.add_argument("--use_recompute", default=None, type=str2bool, help="whether use recompute")
+    parser.add_argument(
+        "--enable_flash_attention",
+        default=None,
```
[the remark this answers is not in the document]
Comment (congw729) [reply examples/stable_diffusion_v2/train_text_to_image.py:177]: will remove the duplicate codes.

**Review by SamitHuang — APPROVED** [review]

**Review by Gaohan123 — APPROVED** [review]

**Review by wtomin — APPROVED** [review]
